### PR TITLE
Fix mm_token_type_ids silently dropped in DPO VLM training

### LINKED ISSUE #5277: DPOTrainer silently ignores mm_token_type_ids when training with Qwen2.5-VL and similar 3D-RoPE models
DPOTrainer silently ignores mm_token_type_ids when training with Qwen2.5-VL and similar 3D-RoPE models.

### Description

Since transformers v5.3.0 (huggingface/transformers#43972), Qwen2_5_VLForConditionalGeneration and similar models using 3D RoPE return mm_token_type_ids from the processor by default. DPOTrainer silently ignores this field throughout training, resulting in incorrect positional encodings for all VLM DPO runs with affected models.

Unlike the analogous bugs in SFT (#5177) and GRPO/RLOO (#5175), which produced hard crashes, DPO fails silently and is therefore not caught by CI.

### Impact

All DPO training runs using Qwen2_5_VLForConditionalGeneration (or any other model that returns mm_token_type_ids) on transformers ≥ 5.3.0 produce wrong positional encodings. Training completes without error, making the bug invisible without inspecting model outputs.

### PR DESCRIPTION
**Pull request by albertvillanova**
Fix `mm_token_type_ids` silently dropped in DPO VLM training.

Fix #5277.

---

> [!NOTE]
> **Medium Risk**
> Touches the VLM batching/forward kwargs path for DPO, so mistakes could break multimodal training or cause subtle tensor-shape/runtime errors, but the change is narrow and covered by a targeted regression test.
> 
> **Overview**
> Fixes DPO vision-language training for processors that return `mm_token_type_ids` (e.g. Qwen2.5-VL), ensuring the field is **not dropped** and its tensor shape stays aligned with `input_ids`.
> 
> `DataCollatorForVisionPreference` now pads/flushes `mm_token_type_ids` alongside other batch tensors and includes it in the batch output, and `DPOTrainer` forwards `mm_token_type_ids` into both reference-logprob computation and the main loss forward pass. Adds a regression test validating `mm_token_type_ids` presence and shape.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 4842af6753c1360eff8ed799050906f2111607de. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: trl/trainer/dpo_trainer.py [hunk 1]
```diff
@@ -1113,7 +1133,14 @@ def _compute_loss(self, model, inputs, return_outputs):
...
         model_kwargs = {"input_ids": input_ids, "attention_mask": attention_mask, "use_cache": False}
-        for key in ("pixel_values", "pixel_attention_mask", "image_grid_thw", "image_sizes", "token_type_ids"):
+        for key in (
+            "pixel_values",
+            "pixel_attention_mask",
+            "image_grid_thw",
+            "image_sizes",
+            "token_type_ids",
+            "mm_token_type_ids",
+        ):
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/dpo_trainer.py:1143]: yep, it sounds like a legit feedback

[the remark this answers is not in the document]
Comment (albertvillanova) [reply trl/trainer/dpo_trainer.py:1143]: @qgallouedec,

First, this is a pre-existing condition, not introduced by this PR:
- The same gap for token_type_ids has never caused a filed issue

Second, as already explicitly documented in the tests, for VLMs, truncating can remove image tokens, leading to errors. Anyone doing VLM DPO uses `max_length=None`

Therefore, in my opinion, if this is an issue we want to handle, I think it should be done in separate PR. I am opening one! :rocket:

[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/dpo_trainer.py:1143]: > The same gap for token_type_ids has never caused a filed issue

interesting, curious to know why

> Anyone doing VLM DPO uses max_length=None

yes that's recommended, but careful user might use `max_length!=0` (technically it's supported, if you ensure that no image token are truncated)

ok a separate pr sounds good!
